synapse/lib/httpapi.py

```diff
@@ -293,11 +293,13 @@ async def _reqValidOpts(self, opts):
         if opts is None:
             opts = {}
 
-        user = await self.user()
+        authcell = self.getAuthCell()
+        useriden = await self.useriden()
 
-        opts.setdefault('user', user.iden)
-        if opts.get('user') != user.iden:
-            user.confirm(('impersonate',))
+        opts.setdefault('user', useriden)
+        if opts.get('user') != useriden:
+            if not await authcell.isUserAllowed(useriden, ('impersonate',)):
+                opts['user'] = useriden
```

Comment:
It feels like trying to execute a storm query as a different user than the session usrer, and not having the permissions to impersonate the other person, should fail.  This is actually a change in behavior from the V1 api and differs from the telepath storm() api.